algorithms/src/Kokkos_Sort.hpp

```diff
@@ -665,6 +667,183 @@ void sort(ViewType view, size_t const begin, size_t const end) {
   exec.fence("Kokkos::Sort: fence after sorting");
 }
 
+namespace Impl {
+
+// true for TeamVectorRange, false for ThreadVectorRange
+template <bool teamLevel>
+struct NestedRange {};
+
+// Specialization for team-level
+template <>
+struct NestedRange<true> {
+  template <typename TeamMember, typename SizeType>
+  KOKKOS_FORCEINLINE_FUNCTION static auto create(const TeamMember& t,
+                                                 SizeType len) {
+    return Kokkos::TeamVectorRange(t, len);
+  }
+  template <typename TeamMember>
+  KOKKOS_FORCEINLINE_FUNCTION static void barrier(const TeamMember& t) {
+    t.team_barrier();
+  }
+};
+
+// Specialization for thread-level
+template <>
+struct NestedRange<false> {
+  template <typename TeamMember, typename SizeType>
+  KOKKOS_FORCEINLINE_FUNCTION static auto create(const TeamMember& t,
+                                                 SizeType len) {
+    return Kokkos::ThreadVectorRange(t, len);
+  }
+  // Barrier is no-op, as vector lanes of a thread are implicitly synchronized
+  // after parallel region
+  template <typename TeamMember>
+  KOKKOS_FORCEINLINE_FUNCTION static void barrier(const TeamMember&) {}
+};
+
+// When just doing sort (not sort_by_key), use nullptr_t for ValueViewType.
+// This only takes the NestedRange instance for template arg deduction.
+template <class TeamMember, class KeyViewType, class ValueViewType,
+          class Comparator, bool useTeamLevel>
+KOKKOS_INLINE_FUNCTION void sort_nested_impl(const TeamMember& t,
+                                             const KeyViewType& keyView,
+                                             const ValueViewType& valueView,
+                                             const Comparator& comp,
+                                             const NestedRange<useTeamLevel>&) {
+  using SizeType  = typename KeyViewType::size_type;
+  using KeyType   = typename KeyViewType::non_const_value_type;
+  using Range     = NestedRange<useTeamLevel>;
+  SizeType n      = keyView.extent(0);
+  SizeType npot   = 1;
+  SizeType levels = 0;
+  // FIXME: ceiling power-of-two is a common thing to need - make it a utility
```

Comment:
Adding a reference to #4577 so we know to update when it eventually lands into `develop` (I am not actively working on it)